### PR TITLE
CoH - Ikoma Kiyono

### LINKED ISSUE #3439: CoH - Ikoma Kiyono
image

### PR DESCRIPTION
**Pull request by The-Outcast**
- Implement GloryCount actions and ClaimFavor actions.
- Implement Ikoma Kiyono
- Rework some older cards with the new action

closes #3439

### CODE REVIEW

File: server/game/cards/06-CotE/SeveredFromTheStream.js [hunk 1]
```diff
@@ -5,12 +5,10 @@ class SeveredFromTheStream extends DrawCard {
...
         this.action({
             title: 'Return player\'s rings',
-            gameAction: AbilityDsl.actions.returnRing(context => {
-                if(context.player.opponent && context.player.getGloryCount() < context.player.opponent.getGloryCount()) {
-                    return { target: context.player.getClaimedRings() };
-                } else if(context.player.opponent && context.player.getGloryCount() > context.player.opponent.getGloryCount()) {
-                    return { target: context.player.opponent.getClaimedRings() };
-                }
+            gameAction: AbilityDsl.actions.performGloryCount({
+                gameAction: winner => winner && AbilityDsl.actions.returnRing({
```
Comment (jeremylarner) [inline server/game/cards/06-CotE/SeveredFromTheStream.js]: need to check for an opponent here

File: server/game/GameActions/ClaimFavorAction.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import { GameAction, GameActionProperties } from './GameAction';
+
+export interface ClaimFavorProperties extends PlayerActionProperties {
+    target: Player | null;
+}
+
+export class ClaimFavorAction extends PlayerAction {
+    name = 'discardFavor';
+    eventName = EventNames.OnClaimFavor;
+    effect = '{0} succesfully claims the Emperor\'s favor.';
```
Comment (jeremylarner) [inline server/game/GameActions/ClaimFavorAction.ts]: this `effect` is similar to `effect` in a triggered ability - it should be `'claim the Emperor\'s favor'`

File: server/game/cards/09.4-TCoH/IkomaKiyono.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+class IkomaKiyono extends DrawCard {
+    setupCardAbilities() {
+        this.wouldInterrupt({
+            title: 'Ready for Glory Count',
+            when: {
+                onGloryCount: (event, context) => { // eslint-disable-line no-unused-vars
+                    return context.player && context.player.opponent && context.player.honor > context.player.opponent.honor;
+                }
+            },
+            effect: 'ready for the incoming glory count.',
```
Comment (jeremylarner) [inline server/game/cards/09.4-TCoH/IkomaKiyono.js]: not sure I like 'incoming' - isn't the default fine here?

File: server/game/cards/06-CotE/SeveredFromTheStream.js (hunk 1 above)
Comment (jeremylarner) [reply server/game/cards/06-CotE/SeveredFromTheStream.js]: Also, do we need an `effect:` here, or are we happy with the default

**Comment by jeremylarner** [comment]
Can you remove the changes to the submodule here too?

File: server/game/cards/09.4-TCoH/IkomaKiyono.js (hunk 3 above)
Comment (The-Outcast) [reply server/game/cards/09.4-TCoH/IkomaKiyono.js]: if 'player uses Ikoma Kiyono to ready Ikoma Kiyono is what you want?

File: server/game/GameActions/ClaimFavorAction.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    hasLegalTarget(): boolean {
+        return true;
+    }
+
+    canAffect(player: Player, context: AbilityContext, additionalProperties = {}): boolean {
+        let properties = this.getProperties(context, additionalProperties) as ClaimFavorProperties;
+        let target = properties.target;
+
+        return player && !player.imperialFavor && super.canAffect(player, context);
```
Comment (Matthew-Griffith) [inline server/game/GameActions/ClaimFavorAction.ts]: here if player is undefined this method would not return a boolean. maybe we should say `!!player && ..`

File: server/game/GameActions/ClaimFavorAction.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import AbilityContext = require('../AbilityContext');
+import Player = require('../player');
+import { GameAction, GameActionProperties } from './GameAction';
+
+export interface ClaimFavorProperties extends PlayerActionProperties {
+    target: Player | null;
+}
+
+export class ClaimFavorAction extends PlayerAction {
+    name = 'discardFavor';
```
Comment (jamesoddy) [inline server/game/GameActions/ClaimFavorAction.ts]: Should this be `name = 'claimFavor'` ?  I'm not sure where the action names are used.

**Pull request merged**